lib/versioning/pvp/index.ts

```diff
@@ -0,0 +1,165 @@
+import { VersioningApi } from '../types';
+
+export const id = 'pvp';
+export const displayName = 'PVP';
+export const urls = ['https://pvp.haskell.org'];
+export const supportsRanges = true;
+
+// https://pvp.haskell.org/#version-numbers
+type PVP = {
+  A: number; // Major
+  B: number; // Major
+  C: number; // Minor
+  additional: number[]; // First element is patch. Can be empty (no patch version)
+};
+
+const hasLeadingZero = (str: string): boolean =>
+  str.length > 1 && str.startsWith('0');
+
+const anyElementHasLeadingZero = (strings: string[]): boolean => {
+  const bools = strings.map((str) => hasLeadingZero(str));
+  return bools.includes(true);
+};
+
+const throwInvalidVersionError = (version: string): void => {
+  throw new Error(
+    `${version} is not a valid version. See https://pvp.haskell.org.`
+  );
+};
+
+const parse = (version: string): PVP => {
+  const [a, b, c, ...additional] = version.split('.');
+  if (anyElementHasLeadingZero([a, b, c, additional].flat())) {
+    throwInvalidVersionError(version);
+  }
+  const a_ = parseInt(a);
+  const b_ = parseInt(b);
+  const c_ = parseInt(c);
+  if (isNaN(a_) || isNaN(b_) || isNaN(c_)) {
+    throwInvalidVersionError(version);
+  }
+  const additional_: number[] = [];
+  additional.forEach((ele) => {
+    const ele_ = parseInt(ele);
+    if (isNaN(ele_)) {
+      throwInvalidVersionError(version);
+    } else {
+      additional_.push(ele_);
+    }
+  });
+  return { A: a_, B: b_, C: c_, additional: additional_ };
+};
+
+/**
+ * Represents major version as a float
+ * A.B
+ * B is optional so if it is not included it defaults to A.0
+ */
+const getMajor = (version: string): number => {
+  const { A, B } = parse(version);
+  return parseFloat(`${A}.${B}`);
+};
+
+const getMinor = (version: string): number => {
+  const { C } = parse(version);
+  return C;
+};
+
+// Patch version is optional
+const getPatch = (version: string): number | null => {
+  const { additional } = parse(version);
+  return additional[0];
+};
+
+const isValid = (version: string): boolean => {
+  try {
+    parse(version);
+  } catch {
+    return false;
+  }
+  return true;
+};
+
+/**
+ * PVP does not specify a stable version
+ * https://pvp.haskell.org/faq/#how-does-the-pvp-relate-to-semantic-versioning-semver
+ */
+const isStable = (): boolean => true;
+
+const isVersion = (version: string): string | boolean | null =>
+  !!isValid(version);
+
+// credit: https://stackoverflow.com/a/22015930/12963115
```

Comment:
I don't know if it is safe legally to pull in this section of the code?
The code we have here looks similar to the code that you can see if you follow the StackOverflow link.

From reading the StackOverflow Terms of service licensing section [^stackoverflow], it seems that all contributions are licensed under the Attribution-ShareAlike 4.0 International (CC BY-SA 4.0) license [^creative-commons].

[^stackoverflow]: https://stackoverflow.com/legal/terms-of-service/public#licensing
[^creative-commons]: https://creativecommons.org/licenses/by-sa/4.0/

Reply:
CC BY-SA 4.0 is compatible with GPLv3 therefore AGPLv3 https://en.m.wikipedia.org/wiki/License_compatibility#CC_BY-SA_and_GPLv3

Reply:
I've removed this code now